diff --git a/go.mod b/go.mod
index cc578900..464088ad 100644
--- a/go.mod
+++ b/go.mod
@@ -1,6 +1,6 @@
 module github.com/ollama/ollama
 
-go 1.24.0
+go 1.23.0
 
 require (
 	github.com/containerd/console v1.0.3
diff --git a/server/create.go b/server/create.go
index 4294554b..44169f72 100644
--- a/server/create.go
+++ b/server/create.go
@@ -230,21 +230,11 @@ func convertFromSafetensors(files map[string]string, baseLayers []*layerGGML, is
 		return nil, err
 	}
 	defer os.RemoveAll(tmpDir)
-	// Set up a root to validate paths
-	root, err := os.OpenRoot(tmpDir)
-	if err != nil {
-		return nil, err
-	}
-	defer root.Close()
 
 	for fp, digest := range files {
 		if !fs.ValidPath(fp) {
 			return nil, fmt.Errorf("%w: %s", errFilePath, fp)
 		}
-		if _, err := root.Stat(fp); err != nil && !errors.Is(err, fs.ErrNotExist) {
-			// Path is likely outside the root
-			return nil, fmt.Errorf("%w: %s: %s", errFilePath, err, fp)
-		}
 
 		blobPath, err := GetBlobsPath(digest)
 		if err != nil {
